lib/python/numpy.h

```diff
@@ -85,54 +76,96 @@ auto cast_to_array_like(const py::object &obj, const units::Unit unit) {
-      core::parallel::blocked_range(0, r.shape(0)), [&](const auto &range) {
-        auto it = begin + range.begin() * r.shape(1);
-        for (scipp::index i = range.begin(); i < range.end(); ++i)
-          for (scipp::index j = 0; j < r.shape(1); ++j, ++it)
-            copy_element<convert>(r(i, j), *it);
-      });
+template <bool convert, class T, class Dst>
+void copy_flattened_0d(const typed_buffer<T> &src, Dst &dst) {
+  copy_element<convert>(*src.ptr, *dst);
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_3d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  for (scipp::index i = 0; i < r.shape(0); ++i)
-    for (scipp::index j = 0; j < r.shape(1); ++j)
-      for (scipp::index k = 0; k < r.shape(2); ++k, ++it)
-        copy_element<convert>(r(i, j, k), *it);
+template <bool convert, class T, class Dst>
+void copy_flattened_inner_dim(const typed_buffer<T> &src, Dst &dst,
+                              ssize_t length, const ssize_t dim,
+                              const ssize_t offset) {
+  const auto stride = src.stride(dim);
+  length = length == -1 ? src.shape[dim] : length;
+  for (scipp::index i = 0; i < length; ++i) {
+    copy_element<convert>(src.ptr[i * stride + offset], *dst);
+    ++dst;
+  }
 }
 
-template <bool convert, class T, class View>
-void copy_flattened_4d(const py::array_t<T> &data, View &&view) {
-  auto r = data.unchecked();
-  auto it = view.begin();
-  for (scipp::index i = 0; i < r.shape(0); ++i)
-    for (scipp::index j = 0; j < r.shape(1); ++j)
-      for (scipp::index k = 0; k < r.shape(2); ++k)
-        for (scipp::index l = 0; l < r.shape(3); ++l, ++it)
-          copy_element<convert>(r(i, j, k, l), *it);
+template <bool convert, class T, class Dst>
+void copy_flattened_middle_dims(const typed_buffer<T> &src, Dst &dst,
+                                ssize_t length, const ssize_t dim,
+                                const ssize_t offset) {
+  if (dim + 1 == src.ndim)
+    copy_flattened_inner_dim<convert>(src, dst, length, dim, offset);
+  else {
+    const auto stride = src.stride(dim);
+    length = length == -1 ? src.shape[dim] : length;
+    for (scipp::index i = 0; i < length; ++i)
+      copy_flattened_middle_dims<convert>(src, dst, -1, dim + 1,
+                                          i * stride + offset);
+  }
+}
+
+template <class T> ssize_t inner_volume(const typed_buffer<T> &buffer) {
+  return std::accumulate(std::next(buffer.shape.begin()), buffer.shape.end(), 1,
+                         std::multiplies<ssize_t>{});
+}
+
+template <bool convert, class T, class Dst>
+void copy_flattened(const typed_buffer<T> &src, Dst &dst) {
+  if (src.ndim == 0)
+    copy_flattened_0d<convert>(src, dst);
+  else {
+    const auto src_stride = src.stride(0);
+    const auto dst_stride = inner_volume(src);
+    core::parallel::parallel_for(
+        core::parallel::blocked_range(0, src.shape[0]), [&](const auto &range) {
```

Comment:
It seems this approach will be very suboptimal if we either have `shape[0]==1` (or very small), or `shape[-1]==1` (or very small). Can you benchmark those two cases as well? In particular the 2-D cases, i.e., (N, 1) and (1, N). The old implementation likely had a similar problem. I wonder if one could `squeeze` such dims?

Or maybe we can move the `parallel_for` call into `copy_flattened_middle_dims`, and make it conditional based on the current dim's size, and whether any out dim was processed in parallel already?